deps/sketch.py

```diff
@@ -0,0 +1,271 @@
+
+
+async def _compile_sketch(
+    sketch: Sketch, installed_libs: dict[Library, str]
+) -> dict[str, str]:
+    async with aiofiles.tempfile.TemporaryDirectory() as dir_name:
+        sketch_path = f"{dir_name}/src/main.cpp"
+        platformio_config_path = f"{dir_name}/platformio.ini"
+
+        os.mkdir(f"{dir_name}/src")
+
+        # Write the sketch to a temp .ino file
+        async with aiofiles.open(sketch_path, "w+") as _f:
+            await _f.write("#include <Arduino.h>\n" + sketch.source_code)
+
+        async with aiofiles.open(platformio_config_path, "w+") as _f:
+            libs = ""
+            includes = ""
+            for lib in installed_libs:
+                libs += f"\n\t\t\t{CWD}/arduino-libs/{lib}@{installed_libs[lib]}/src "
+                includes += f"-I'{CWD}/arduino-libs/{lib}@{installed_libs[lib]}/src' "
+                async with aiofiles.open(
+                    f"{CWD}/arduino-libs/{lib}@{installed_libs[lib]}/compiled_sources.json",
+                    "r",
+                ) as _f2:
+                    data = json.loads(await _f2.read())
+                    includes += data["include"][fqbn_to_board[sketch.board]].replace(
+                        "../", f"{CWD}/arduino-libs/"
+                    )
+                    libs += "\n" + data["dirs"][fqbn_to_board[sketch.board]].replace(
+                        "../", f"{CWD}/arduino-libs/"
+                    )
+            await _f.write(
+                f"[env:build]\nplatform = {fqbn_to_platform[sketch.board]}\nbuild_flags = -w {includes}\nboard = {fqbn_to_board[sketch.board]}\nframework = arduino\nlib_deps = {libs}"  # pylint: disable=line-too-long
+            )
+
+        compiler = await asyncio.create_subprocess_exec(
+            "platformio",
+            "run",
+            stderr=asyncio.subprocess.PIPE,
+            stdout=asyncio.subprocess.PIPE,
+            cwd=dir_name,
+        )
+        stdout, stderr = await compiler.communicate()
+        if compiler.returncode != 0:
+            logger.warning("Compilation failed: %s", stderr.decode() + stdout.decode())
+            raise HTTPException(500, stderr.decode() + stdout.decode())
+
+        output_file = f"{dir_name}/.pio/build/build/firmware.hex"
+        async with aiofiles.open(output_file, "r", encoding="utf-8") as _f:
+            return {"hex": str(await _f.read())}
+
+
+@asynccontextmanager
+async def startup(_app: FastAPI) -> None:
+    """Startup context manager"""
+    if settings.library_index_refresh_interval > 0:
+        await refresh_library_index()
+    yield
+
+
+@repeat_every(seconds=settings.library_index_refresh_interval, logger=logger)
+async def refresh_library_index():
+    """Update the Arduino library index"""
+    if not await check_for_internet():
+        return
+    logger.info("Updating library index...")
+    global library_index_json, library_indexed_json  # pylint: disable=global-statement
+    async with httpx.AsyncClient() as client:
```

Comment:
if this becomes a class, you can store the httpx client an re-use it instead of re-creating it for every request